citation-matching/citation-matching-core-code/src/main/java/pl/edu/icm/coansys/citations/CitationMatchingJob.java

```diff
@@ -80,60 +53,69 @@ public static void main(String[] args) throws IOException, InstantiationExceptio
         
+        InputCitationConverter inputCitationConverter = Class.forName(params.inputCitationConverterClass).asSubclass(InputCitationConverter.class).newInstance();
+        citationMatchingService.setInputCitationConverter(inputCitationConverter);
         
-        return entities;
     }
-    
-    private static void saveMatchedCitations(JavaSparkContext sc, JavaPairRDD<MatchableEntity, IdWithSimilarity> matchedCitations, String outputDirPath) {
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    private static void createAndSetInputDocumentReaderAndConverter(ConfigurableCitationMatchingService service, JavaSparkContext sc, CitationMatchingJobParameters params) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
         
-        matchedCitations
-            .mapToPair(x -> new Tuple2<TextWithBytesWritable, Text>(
-                    new TextWithBytesWritable(x._1.id(), x._1.data().toByteArray()),
-                    new Text(x._2.getSimilarity() + ":" + x._2.getId())))
-            .saveAsNewAPIHadoopFile(outputDirPath, TextWithBytesWritable.class, Text.class, SequenceFileOutputFormat.class);
+        InputDocumentReader inputDocumentReader = Class.forName(params.inputDocumentReaderClass).asSubclass(InputDocumentReader.class).newInstance();
+        inputDocumentReader.setSparkContext(sc);
+        service.setInputDocumentReader(inputDocumentReader);
+        
+        InputDocumentConverter inputDocumentConverter = Class.forName(params.inputDocumentConverterClass).asSubclass(InputDocumentConverter.class).newInstance();
+        service.setInputDocumentConverter(inputDocumentConverter);
+        
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    private static void createAndSetOutputWriterAndConverter(ConfigurableCitationMatchingService citationMatchingService, JavaSparkContext sc, CitationMatchingJobParameters params) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
+        
+        OutputWriter outputWriter = Class.forName(params.outputWriterClass).asSubclass(OutputWriter.class).newInstance();
+        citationMatchingService.setOutputWriter(outputWriter);
+        
+        OutputConverter outputConverter = Class.forName(params.outputConverterClass).asSubclass(OutputConverter.class).newInstance();
+        citationMatchingService.setOutputConverter(outputConverter);
         
     }
     
     
-    private static JavaPairRDD<String, String> matchCitDocByHashes(JavaSparkContext sc,
-            JavaPairRDD<String, MatchableEntity> citations, JavaPairRDD<String, MatchableEntity> documents,
-            List<Pair<MatchableEntityHasher, MatchableEntityHasher>> entitiesHashers, long maxHashBucketSize) {
+    private static CoreCitationMatchingService createCoreCitationMatchingService(JavaSparkContext sc, CitationMatchingJobParameters params) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
```

Comment:
I think it would be a nice idea to move the creation of CoreCitationMatchingService to a separate factory, e.g. CoreCitationMatchingSimpleFactory. This way we'll be able to test the creation of CCMService and easily create it in external jobs.